### PR TITLE
JsonApiDocument hashCode and equals were inconsistent

### PR DESCRIPTION
**Pull request by wcekan**
## Description
Pulled out only the code changes from #1162 for easier review.
- JsonApiDocument hashCode and equals were inconsistent.
- Correctly process empty ConstraintViolations
- Remove unnecessary initializations in code

## Motivation and Context
Stability

## How Has This Been Tested?
Unit testing

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java [hunk 1]
```diff
@@ -326,5 +339,21 @@ public void readListWithMeta() throws IOException {
...
+
+    private void checkEquality(JsonApiDocument doc1) {
+        JsonApiDocument doc2;
+        try {
+            String json = mapper.writeJsonApiDocument(doc1);
+            doc2 = mapper.readJsonApiDocument(json);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        if (doc2.getData() == null) {
```
Comment (aklish) [inline elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: Why do we do this?

Comment (wcekan) [reply elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: Good question.  Seems the null is not handled well.

I found the problem. Somehow serializing changes the value of doc.data!

Comment (wcekan) [reply elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: I removed this hack and added a verification that data is unchanged after serialization, which fails :(

Comment (wcekan) [reply elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: It seems "Empty" object and Null object are written the same, but read differently.

Comment (wcekan) [reply elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: expected: 
<JsonApiDocument(data=Data(values=[], relationshipType=MANY_TO_ONE), meta=null, links=null, includedRecs=[], included=null)> 
but was: 
<JsonApiDocument(data=null, meta=null, links=null, includedRecs=[], included=null)>

Comment (wcekan) [reply elide-core/src/test/java/com/yahoo/elide/jsonapi/JsonApiTest.java]: Will now treat "Empty" object and Null data object as equal.

**Pull request merged**